### PR TITLE
README.aipl to generate the README.md

### PR DESCRIPTION
**Pull request by saulpw**

### CODE REVIEW

File: README.aipl [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+    aipl.operators.values()
+ !format
+    - {op.__name__} (in={op.rankin} out={op.rankout}): {op.__doc__}
+ !join sep=\n
+
+!!def read-summary
+  !read chains/summarize.aipl
+
+
+!!read-summary>summarize_aipl
```
Comment (anjakefala) [inline README.aipl:12]: With these `!!`, `summarize_aipl` is added to the global context, with the contents of `!read chains/summarize.sipl`. Then, you can reference it later in the pipeline via `summarize_aipl`.

File: aipl/interpreter.py [hunk 2]
```diff
@@ -116,8 +116,12 @@ def set_last_prompt(ret, prompt):
...
                 prompt = ''
                 if ret and ret[-1].immediate:  # !!op means do the command immediately
-                    result = self.eval_op(ret[-1], Table(), contexts=[self.globals])
-                    stderr('compile time:', result)
+                    cmd = ret[-1]
+                    result = self.eval_op(cmd, Table(), contexts=[self.globals])
+                    if cmd.varnames:
+                        k = cmd.varnames[-1]
+                        self.globals[k] = result
+                        stderr(f'(global) {k} = {cmd.opname}:', result)
```
Comment (anjakefala) [inline aipl/interpreter.py:124]: This adds the results of `!!` operations to the global context.

File: aipl/interpreter.py [hunk 3]
```diff
@@ -287,10 +291,26 @@ def prep_output(aipl,
...
+                raise Exception(f'unknown type for in_row: {type(in_row)}')
+
+            rows = []
+            for v in out:
+                d = {'__parent': parent_row} if parent_row is not None else {}
+                d[varname] = v
+                rows.append(d)
+            ret = Table(rows, parent=parent_table)
+#            for k in ret.rows[0].keys():  # assumes first row has same keys as every other row
+#                ret.add_column(Column(k))
```
Comment (anjakefala) [inline aipl/interpreter.py:313]: @saulpw do you remember why you made this change?

File: aipl/ops/def.py [hunk 4]
```diff
@@ -18,9 +18,15 @@ def op_def(aipl, opname, prompt=''):
            rankin=cmds[0].op.rankin,
            rankout=cmds[-1].op.rankout,
            arity=cmds[0].op.arity)
-    def new_operator(aipl, input, *args, **kwargs):
-        argkey = aipl.unique_key
-        ret = aipl.run_cmdlist(cmds, Table([{argkey:input}]), *args)
+    def new_operator(aipl, *args, **kwargs):
+        arity = cmds[0].op.arity
```
Comment (anjakefala) [inline aipl/ops/def.py:22]: Handles `!!` operators that call commands with a 0 arity.

File: aipl/ops/op_text.py [hunk 5]
```diff
@@ -1,12 +1,14 @@
...
+
 from aipl import defop
 from aipl.table import LazyRow, Table, Column
 
 
 @defop('format', 0.5, 0)
 def op_format(aipl, row:LazyRow, prompt:str='') -> str:
     'Format prompt text as template, substituting values from row'
-    return prompt.format_map(row)
+    return prompt.format_map(ChainMap(aipl.globals, row))
```
Comment (anjakefala) [inline aipl/ops/op_text.py:11]: Passes the global context to `!format`, so we can reference `!!` outputs.

**Review by anjakefala — APPROVED** [review]

File: aipl/interpreter.py (hunk 3 above)
Comment (saulpw) [reply aipl/interpreter.py:313]: The `in_row` was None for the arity=0, rankout=1.5 case for `!python-input`.

**Pull request merged**